lib/phase/functional.js

```diff
@@ -0,0 +1,149 @@
+'use strict';
+
+const Joi = require('joi');
+const Hoek = require('hoek');
+const clone = require('clone');
+
+// Actual environment size is limited by space, not quantity.
+// We do this to force sd.yaml to be simpler.
+const MAX_ENVIRONMENT_VARS = 25;
+// If they are trying to execute >25 permutations,
+// maybe there is a better way to accomplish their goal.
+const MAX_PERMUTATIONS = 25;
+
+/**
+ * Make sure the Matrix they specified is valid
+ * @method validateJobMatrix
+ * @param  {Object}      job     Job to inspect
+ * @param  {String}      jobName Name of job (for error reporting)
+ * @return {Array}               List of errors
+ */
+function validateJobMatrix(job, jobName) {
+    let matrixSize = 1;
+    const errors = [];
+    const matrix = Hoek.reach(job, 'matrix', {
+        default: {}
+    });
+
+    Object.keys(matrix).forEach((row) => {
+        matrixSize *= matrix[row].length;
+    });
+
+    if (matrixSize > MAX_PERMUTATIONS) {
+        errors.push(`Job ${jobName}: Matrix cannot contain >${MAX_PERMUTATIONS} permutations`);
+    }
+
+    return errors;
+}
+
+/**
+ * Ensure the job is something that can be run
+ *  - has at least one step
+ *  - the image name
+ *  - not too many environment variables
+ * @method validateJobSchema
+ * @param  {Object}          doc Document that went through flattening
+ * @return {Array}               List of errors
+ */
+function validateJobSchema(doc) {
+    let errors = [];
+
+    // Jobs
+    const SCHEMA_JOB = Joi.object()
+        .keys({
+            steps: Joi.object().min(1),
```

Comment:
min(1) only validates that if `steps` exists, then it has to have at least 1 key. If `steps` doesn't exist then this will still pass